File: leetcode/cpp/16.3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target)
    {
        sort(begin(nums), end(nums));
        int res = 0x00ffffff;
        for (int i = 0; i < nums.size() - 2; ++i) {
            int j = i + 1, k = nums.size() - 1;
            if (j >= nums.size() - 1)
                break;
            while (j < k) {
                int t = nums[i] + nums[j] + nums[k];
                if (abs(t - target) < abs(res - target)) {
                    res = nums[i] + nums[j] + nums[k];
                }
                if (t > target) {
                    --k;
                } else {
                    ++j;
                }
            }
        }

        return res;
    }
};

File: leetcode/python/16.3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        res = nums[0] + nums[1] + nums[2]

        for i in range(len(nums)-2):
            l = i + 1
            r = len(nums) - 1
            while l < r:
                t = nums[i] + nums[l] + nums[r]

                if abs(target - t) < abs(target - res):
                    res = t
                if target == t:
                    return res
                if t > target:
                    r -= 1

                else:
                    l += 1
        return res
